vortex-scan/src/multi_scan.rs

```diff
@@ -12,27 +12,29 @@ use vortex_error::VortexResult;
 
 use crate::ScanBuilder;
 
-type ArrayFuture = BoxFuture<'static, VortexResult<Option<ArrayRef>>>;
-type ScanBuilderFactory = Arc<SegQueue<Box<dyn FnOnce() -> ScanBuilder<ArrayRef> + Send + Sync>>>;
+type ArrayFuture<S> = BoxFuture<'static, VortexResult<Option<(ArrayRef, S)>>>;
+type ScanBuilderFactory<S> =
+    Arc<SegQueue<Box<(dyn FnOnce() -> ScanBuilder<(ArrayRef, S)> + Send + Sync)>>>;
 
 /// Coordinator to orchestrate multiple scan operations.
 ///
 /// `MultiScan` allows to queue multiple scan operations in order to execute
 /// them in parallel. In particular, this enables scanning multiple files.
-#[derive(Default)]
-pub struct MultiScan {
-    scan_builder_factory: ScanBuilderFactory,
+pub struct MultiScan<S> {
+    scan_builder_factory: ScanBuilderFactory<S>,
 }
 
-impl MultiScan {
+impl<S> MultiScan<S> {
     pub fn new() -> Self {
-        Self::default()
+        Self {
+            scan_builder_factory: Arc::new(SegQueue::new()),
+        }
     }
 
-    /// `ScanBuilder`s are passed through closures to decouple how the they are created.
+    /// Add lazily constructed scan builders paired with their corresponding states.
     pub fn with_scan_builders<I, F>(self, closures: I) -> Self
     where
-        F: FnOnce() -> ScanBuilder<ArrayRef> + 'static + Send + Sync,
+        F: FnOnce() -> ScanBuilder<(ArrayRef, S)> + 'static + Send + Sync,
```

Comment:
You have the caller in vortex-duckdb map the scan inside the closure. So you don't need to enforce that this is a tuple anymore, you can just take an untyped `A` same as regular scan?